File: packages/backend/src/routes/tasks.ts
import express from 'express';
import { prisma } from '../lib/prisma.js';
import { TaskStatus } from '@prisma/client';

const router = express.Router();

// GET /api/v1/tasks - List all tasks
router.get('/', async (req, res) => {
  try {
    const { transactionId, status, assignedTo } = req.query;

    const where: any = {};
    if (transactionId) where.transactionId = transactionId as string;
    if (status) where.status = status as TaskStatus;
    if (assignedTo) where.assignedTo = assignedTo as string;

    const tasks = await prisma.task.findMany({
      where,
      orderBy: {
        createdAt: 'asc',
      },
    });

    res.json(tasks);
  } catch (error) {
    console.error('Error fetching tasks:', error);
    res.status(500).json({ error: 'Failed to fetch tasks' });
  }
});

// GET /api/v1/tasks/:id - Get task by ID
router.get('/:id', async (req, res) => {
  try {
    const task = await prisma.task.findUnique({
      where: { id: req.params.id },
      include: {
        transaction: true,
      },
    });

    if (!task) {
      return res.status(404).json({ error: 'Task not found' });
    }

    res.json(task);
  } catch (error) {
    console.error('Error fetching task:', error);
    res.status(500).json({ error: 'Failed to fetch task' });
  }
});

// POST /api/v1/tasks - Create a new task
router.post('/', async (req, res) => {
  try {
    const {
      transactionId,
      title,
      description,
      taskType,
      assignedTo,
      assignedBy = 'system',
      priority = 'NORMAL',
      dueDate,
      isBlocking = false,
      relatedStage,
    } = req.body;

    // Generate task ID
    const year = new Date().getFullYear();
    const count = await prisma.task.count();
    const id = `TASK-${year}-${String(count + 1).padStart(4, '0')}`;

    const task = await prisma.task.create({
      data: {
        id,
        transactionId,
        title,
        description,
        taskType,
        assignedTo,
        assignedBy,
        priority,
        dueDate: dueDate ? new Date(dueDate) : undefined,
        isBlocking,
        relatedStage,
      },
    });

    res.status(201).json(task);
  } catch (error) {
    console.error('Error creating task:', error);
    res.status(500).json({ error: 'Failed to create task' });
  }
});

// POST /api/v1/tasks/:id/complete - Mark task as complete
router.post('/:id/complete', async (req, res) => {
  try {
    const { completedBy, completionNotes, completionData } = req.body;

    const task = await prisma.task.findUnique({
      where: { id: req.params.id },
    });

    if (!task) {
      return res.status(404).json({ error: 'Task not found' });
    }

    if (task.status === TaskStatus.COMPLETED) {
      return res.status(400).json({ error: 'Task is already completed' });
    }

    const updatedTask = await prisma.task.update({
      where: { id: req.params.id },
      data: {
        status: TaskStatus.COMPLETED,
        completedAt: new Date(),
        completedBy,
        completionNotes,
        completionData: completionData || {},
      },
    });

    res.json(updatedTask);
  } catch (error) {
    console.error('Error completing task:', error);
    res.status(500).json({ error: 'Failed to complete task' });
  }
});

// PATCH /api/v1/tasks/:id - Update task
router.patch('/:id', async (req, res) => {
  try {
    const updates = req.body;

    // Convert date strings to Date objects
    if (updates.dueDate) updates.dueDate = new Date(updates.dueDate);
    if (updates.startedAt) updates.startedAt = new Date(updates.startedAt);

    const task = await prisma.task.update({
      where: { id: req.params.id },
      data: updates,
    });

    res.json(task);
  } catch (error) {
    console.error('Error updating task:', error);
    res.status(500).json({ error: 'Failed to update task' });
  }
});

// DELETE /api/v1/tasks/:id - Delete task
router.delete('/:id', async (req, res) => {
  try {
    await prisma.task.delete({
      where: { id: req.params.id },
    });
    res.json({ message: 'Task deleted successfully' });
  } catch (error) {
    console.error('Error deleting task:', error);
    res.status(500).json({ error: 'Failed to delete task' });
  }
});

export default router;

File: legacy/backend/api/tasks.py
"""
Task API Endpoints

Handles task operations - viewing, updating status, completing tasks.
"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import Optional
from pydantic import BaseModel, Field
import logging
from datetime import datetime

from backend.database import get_db
from backend.models import Task, TaskStatus

logger = logging.getLogger(__name__)
router = APIRouter()


class TaskComplete(BaseModel):
    """Schema for completing a task."""
    completion_notes: Optional[str] = Field(None, description="Notes about completion")
    completion_data: Optional[dict] = Field(None, description="Additional completion data")


@router.get("/transactions/{transaction_id}/tasks", response_model=dict)
async def get_transaction_tasks(
    transaction_id: str,
    status_filter: Optional[str] = None,
    is_blocking: Optional[bool] = None,
    db: Session = Depends(get_db)
):
    """
    Get all tasks for a transaction.

    **Logging**: Logs task query for audit trail.
    """
    logger.info(f"Fetching tasks for transaction: {transaction_id}")

    query = db.query(Task).filter(Task.transaction_id == transaction_id)

    if status_filter:
        try:
            status_enum = TaskStatus(status_filter)
            query = query.filter(Task.status == status_enum)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid status: {status_filter}"
            )

    if is_blocking is not None:
        query = query.filter(Task.is_blocking == is_blocking)

    tasks = query.all()

    logger.info(f"Found {len(tasks)} tasks")

    return {
        "transaction_id": transaction_id,
        "count": len(tasks),
        "tasks": [task.to_dict() for task in tasks]
    }


@router.post("/tasks/{task_id}/complete", response_model=dict)
async def complete_task(
    task_id: str,
    completion: TaskComplete,
    db: Session = Depends(get_db)
):
    """
    Mark a task as completed.

    **Workflow Impact**: Completing blocking tasks may allow stage advancement.

    **Logging**: Records task completion with timestamp.
    """
    logger.info("=" * 60)
    logger.info(f"COMPLETING TASK - {task_id}")
    logger.info("=" * 60)

    task = db.query(Task).filter(Task.id == task_id).first()

    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Task {task_id} not found"
        )

    if task.status == TaskStatus.COMPLETED:
        logger.warning(f"Task already completed: {task_id}")
        return {
            "success": True,
            "message": "Task was already completed",
            "task_id": task_id
        }

    logger.info(f"Task: {task.title}")
    logger.info(f"Was blocking: {task.is_blocking}")

    # Update task
    task.status = TaskStatus.COMPLETED
    task.completed_at = datetime.utcnow()
    task.completion_notes = completion.completion_notes
    if completion.completion_data:
        task.completion_data = completion.completion_data

    db.commit()
    db.refresh(task)

    logger.info(f"✓ Task completed at: {task.completed_at}")
    logger.info("=" * 60)

    return {
        "success": True,
        "message": "Task completed successfully",
        "task_id": task_id,
        "completed_at": task.completed_at.isoformat(),
        "was_blocking": task.is_blocking
    }


@router.get("/tasks/{task_id}", response_model=dict)
async def get_task(
    task_id: str,
    db: Session = Depends(get_db)
):
    """Get details of a specific task."""
    task = db.query(Task).filter(Task.id == task_id).first()

    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Task {task_id} not found"
        )

    return task.to_dict()
